[WIP] Histogram support addition to distplot.py

imperorrp opened the pull request:
This PR aims to add support for histograms in `distplot.py` (Issue #24 ), allowing marginal densities to be visualized in this new form, apart from preexisting KDE and ECDF forms



<!-- readthedocs-preview arviz-plots start -->
----
📚 Documentation preview 📚: https://arviz-plots--47.org.readthedocs.build/en/47/

<!-- readthedocs-preview arviz-plots end -->

imperorrp commented:
@OriolAbril I had to dig a bit deeper into understanding how the plotting actually works as coded, but I think I've managed to get step 1 (computing the histogram) roughly done- I'll probably have to combine these DataArrays into a Dataset and then create the visual element for mapping. Also, I'll remove the `print` statements I've added in later commits- just added them to help with seeing the outputs.  Is something like this alright?- Looping through the data variables in `distribution` to get DataArrays and then passing that plus the dimensions to be reduced to the xarray_einstats.histogram() function. 

Also, I was going through the flow of how density was computed for KDE as an example (in Arviz-Stats). I saw bandwidth computing was done for KDE- but that won't be necessary for plotting histograms, right?

imperorrp commented:
Combined the `xarray_einstats.numba.histogram` produced DataArrays into a Dataset: 

![Dataset-image](https://github.com/arviz-devs/arviz-plots/assets/24991131/34a44259-5c4e-44b8-b320-0f0995f96466)

(This is with the Schools dataset) 

It's in the style of the KDE dataset returned by `distribution.azstats.kde` to density

imperorrp commented:
Added a visual element function, backend interface and matplotlib implementation with `matplotlib.pyplot.bar`:

![image](https://github.com/arviz-devs/arviz-plots/assets/24991131/a51a0a3c-73f6-46f8-8f75-35bcea7d2fe9)

The bar widths are different due to the different scales of the x-axis. I'm not sure how this can be fixed yet if keeping the width a consistent value is desired though. Also, the number of bins is set to 20 by default. Some bins have too few values to be visible as their respective bars are too low in height compared to the taller ones.

OriolAbril commented on src/arviz_plots/plots/distplot.py:
```diff
@@ -250,6 +257,33 @@ def plot_dist(
+
+                # number of bins is set to 20 by default
+                hist = histogram(
+                    da=var_data, dims=density_dims, bins=20, **stats_kwargs.get("density", {})
```
the default number of bins should be taken care of by `histogram` function. Moreover, if we believed a different default were needed, it would have to happen via `stats_kwargs` and `setdefault` so a user could modify that value.

imperorrp replied:
Just modified this in the latest commit to do so 

Now by default, with no bins provided (`azp.plot_dist(schools, kind='hist')`) and the same example dataset, this is produced: 
![distplot-hist-matplotlib](https://github.com/arviz-devs/arviz-plots/assets/24991131/84721a96-e8cc-445a-80de-29a9add1175a)

And with bins=40 provided by a user (`azp.plot_dist(schools, kind='hist', stats_kwargs={"density": {"bins":40}})`), this is produced:
![distplot-hist-matplotlib](https://github.com/arviz-devs/arviz-plots/assets/24991131/924f3e90-604b-4b19-b647-b15a3f72d5f7)

Will look at the other comments and make changes too now

OriolAbril commented on src/arviz_plots/backend/__init__.py:
```diff
@@ -99,6 +99,29 @@ def create_plotting_grid(
 
 
 # "geoms"
+def hist(x, height, target, *, width=1.6, bottom=None, align="center", color=None):
```
This will require some changes to the data restructure utility function but I am not too worried about this, at some point we'll have a function in arviz-stats to take care of all this.

OriolAbril commented on src/arviz_plots/backend/__init__.py:
```diff
@@ -99,6 +99,29 @@ def create_plotting_grid(
+def hist(x, height, target, *, width=1.6, bottom=None, align="center", color=None):
+    """Interface for a hist plot.
+
+    Parameters
```
We haven't been completely consistent adding the docstrings to this module, but the idea is for the documentation here so define the behaviour **all** backends have to comply with. It also isn't a general plotting backend interface but a common interface _for our own plots to use_ which means we can be quite restrictive about the features.

For example, I would only accept array_like for x, height, width... I don't think the extra edge case of floats would be used anywhere. It also means the descriptions should not depend on any backend.

OriolAbril commented on src/arviz_plots/backend/__init__.py:
```diff
@@ -99,6 +99,29 @@ def create_plotting_grid(
+        The y coordinate(s) of the bottom side(s) of the bars.
+    align : {'center', 'edge'}, default 'center'
+        Alignment of the bars to the x coordinates.
+    color : color or array-like, optional
+        The colors of the bar faces.
```
I might start with the same color for all rectangles, mostly because I am not sure how easy it will be to support the different color rectangles in all backends. If we then see it is feasible and of interest we can add that extra case.

There should also be a facecolor and an edgecolor arguments given there is a color argument and the generated form is "fillable" to match the expectations from https://arviz-plots.readthedocs.io/en/latest/api/backend/index.html#backend-interface-arguments

OriolAbril reviewed (commented):
It is looking very good already. The computation of the histogram is a bit trickier and would not be part of the gsoc project, but I think it will be helpful to get you more familiar with xarray.

I have tried to focus on the plotting and contributing workflow with the comments. Let me know if you have any doubts related to testing.

imperorrp commented:
> It is looking very good already. The computation of the histogram is a bit trickier and would not be part of the gsoc project, but I think it will be helpful to get you more familiar with xarray.
> 
> I have tried to focus on the plotting and contributing workflow with the comments. Let me know if you have any doubts related to testing.

Thank you, I will take a look at your reviews and advice!

imperorrp commented:
@OriolAbril Latest commit has changes incorporating advice from your comments. Modified data restructure utility function to keep left and right edges info now until the matplotlib backend calculates midpoints from these at the end to create the bars, and I've updated the docstring and color/facecolor/edgecolor interface

OriolAbril commented on src/arviz_plots/backend/matplotlib/__init__.py:
```diff
@@ -200,6 +200,22 @@ def _filter_kwargs(kwargs, artist, artist_kws):
+):
+    """Interface to matplotlib for a histogram bar plot."""
+    artist_kws.setdefault("zorder", 2)
+    bins = list(l_e) + [r_e[-1]]
+    midpoints = [(bins[i] + bins[i + 1]) / 2 for i in range(len(bins) - 1)]
+    if color is not unset:
+        if facecolor is unset:
+            facecolor = color
+        if edgecolor is unset:
+            edgecolor = color
+    kwargs = {"bottom": bottom, "color": facecolor, "edgecolor": edgecolor}
+    return target.bar(midpoints, y, **_filter_kwargs(kwargs, None, artist_kws))
```
To minimize computation, I would do the following:

```
widths = np.asarray(r_e) - np.asarray(l_e)
...
return target.bar(l_e, y, width=widths, align="edge", _**filter_...)
```

The bin edges of a histogram don't necessarly need to be equispaced, so we do need to use array-like widths. We don't need to compute the midpoints though, using `align="edge"` we can use the left edges which we already have as x values.

imperorrp replied:
Okay, I'll add this change and push

imperorrp replied:
This one definitely works better, the gaps between histogram bars seem eliminated now 

![image](https://github.com/arviz-devs/arviz-plots/assets/24991131/5805a397-1b8e-45ea-8a90-a9bf63f39ac8)



![image](https://github.com/arviz-devs/arviz-plots/assets/24991131/4303c7da-dd71-45d1-986e-155746bc08e6)

OriolAbril reviewed (commented):
Left a comment on how to use matplotlib bar function to generate histogram looking plot. Let me know if you need any help to generate the bokeh equivalent

OriolAbril commented:
By the way, I commented on the ess and mcse issues but forgot to comment here. There is a histogram function available in `arviz-stats` now. I think the output will have a similar (if not the same) shape to what you are using

imperorrp commented:
> By the way, I commented on the ess and mcse issues but forgot to comment here. There is a histogram function available in `arviz-stats` now. I think the output will have a similar (if not the same) shape to what you are using

Altered the code to make use of this function. Also, it seems like point estimate and credible intervals are being plotted so close to the x axis for histogram plots because the default 'y' aesthetic value/s returned by plot collection are pretty low in relation. The maximum y-axis values for the KDE/ECDF plots are way lower so it's not a problem for them. I initially thought there was some issue with this aesthetic mapping when type hist is picked, but it works fine- it's just really small values compared to the histogram heights. Should we mention this in the documentation, asking users to define custom 'y' aesthetic values (in `pc_kwargs`)when using kind=histogram for better visibility? 

For the plot_dist example where one subplot is created for each variable:
```
azp.plot_dist(
        data,
        kind="kde",
        pc_kwargs={
            "cols": ["__variable__"],
            "aes": {"color": ["school"], "y": ["school"]},
            "y": np.linspace(0, 0.06, 9),
        },
        aes_map={
            "kde": ["color"],
            "point_estimate": ["color", "y"],
            "credible_interval": ["y"],
        },
    )
    plt.savefig("distplot-kde-cols-matplotlib.png")
```
![image](https://github.com/arviz-devs/arviz-plots/assets/24991131/c029002b-f6fe-4964-9949-9c43fbbbf980)

Point estimates when switching from kind="kde" to kind="hist":
```
azp.plot_dist(
        data,
        kind="hist",
        pc_kwargs={
            "cols": ["__variable__"],
            "aes": {"color": ["school"], "y": ["school"]},
            "y": np.linspace(0, 0.06, 9),
        },
        aes_map={
            "kde": ["color"],
            "point_estimate": ["color", "y"],
            "credible_interval": ["y"],
        },
    )
    plt.savefig("distplot-hist-cols-matplotlib.png")
```
![image](https://github.com/arviz-devs/arviz-plots/assets/24991131/3b5c042e-0a78-4162-a539-fc79cce76717)

This is easily fixable with editing the `np.linspace` line though-
```
"y": np.linspace(0, 600, 9),
```
![image](https://github.com/arviz-devs/arviz-plots/assets/24991131/0d2dedc7-5312-448e-9650-5f50173b8114)

We could maybe add this in the documentation so users know to keep this in mind.

OriolAbril commented on src/arviz_plots/plots/distplot.py:
```diff
@@ -250,6 +253,16 @@ def plot_dist(
             )
 
+        elif kind == "hist":
+            density = distribution.azstats.histogram(
+                dims=density_dims, **stats_kwargs.get("density", {})
+            )
```
we should pass `density=True` as default here. This will normalize the histogram so its area is 1, and will therefore generate histograms with heights similar to those of the kde. That will should take care of the non-reusability of the overlayed example when changing hist/kde. For ecdf/dotplots it will probably break too, but I don't think it matters, ecdf always goes 0-1 whereas dotplots are particularly strange, they should not be expected to act as drop-in replacements of either kde nor hist.

imperorrp replied:
Yeah adding that seems to work for the overlay example, without having to alter the `np.linspace` parameters.

OriolAbril commented on src/arviz_plots/plots/distplot.py:
```diff
@@ -97,6 +99,7 @@ def plot_dist(
 
           * "kde" -> passed to :func:`~arviz_plots.visuals.line_xy`
           * "ecdf" -> passed to :func:`~arviz_plots.visuals.ecdf_line`
+          * "hist" -> passed to :func: `~WIP`
```
update this and add the new visual function to https://github.com/arviz-devs/arviz-plots/blob/main/docs/source/api/visuals.rst

OriolAbril commented on src/arviz_plots/plots/distplot.py:
```diff
@@ -250,6 +253,16 @@ def plot_dist(
 
+        elif kind == "hist":
+            density = distribution.azstats.histogram(
+                dims=density_dims, density=True, **stats_kwargs.get("density", {})
```
`density=True` should be added to `stats_kwargs` with `setdefault`. Then, if a user wanted to they can set it to False, keep the axis and use the y axis to look at counts for example. The task itself isn't really important, what is is that it should be possible

OriolAbril commented on src/arviz_plots/plots/distplot.py:
```diff
@@ -352,7 +374,7 @@ def plot_dist(
             **title_kwargs,
         )
-    if (kind == "kde") and (plot_kwargs.get("remove_axis", True) is not False):
+    if (kind in ("kde", "hist")) and (plot_kwargs.get("remove_axis", True) is not False):
```
I would modify this logic. Instead, I would set the value of the `remove_axis` depending on `kind` at the top, right after copying `plot_kwargs`.

```python
if kind in ("hist", "ecdf"):
    plot_kwargs.setdefault("remove_axis", False)
```

I think as a default we'll want to remove the axis for kde and quantile dotplot but keep it for the others.

OriolAbril commented on src/arviz_plots/visuals/__init__.py:
```diff
@@ -13,6 +13,21 @@
 from arviz_base.labels import BaseLabeller
 
 
+def plot_hist(da, target, backend, **kwargs):
```
I would use `hist` only to "reserve" the `plot_xyz` for batteries included functions, otherwise it might get confusing.

OriolAbril commented on src/arviz_plots/backend/__init__.py:
```diff
@@ -99,6 +99,33 @@ def create_plotting_grid(
 
 
 # "geoms"
+def hist(y, l_e, r_e, target, *, bottom=None, color=None, facecolor=None, edgecolor=None):
```
I went over all potential keywords in https://arviz-plots.readthedocs.io/en/latest/api/backend/index.html#backend-interface-arguments and `alpha` should be added here too.

OriolAbril reviewed (commented):
As always, go over all comments before starting to address any single one of them

imperorrp commented:
Just pushed changes per last review

OriolAbril commented on tests/test_plots.py:
```diff
@@ -85,29 +85,32 @@ def datatree_sample(seed=31):
 
-    def test_plot_dist_models(self, datatree, datatree2, backend):
-        pc = plot_dist({"c": datatree, "n": datatree2}, backend=backend)
+    @pytest.mark.parametrize("kind", ["kde", "hist"])
```
```suggestion
    @pytest.mark.parametrize("kind", ["kde"])
```

OriolAbril commented on src/arviz_plots/backend/bokeh/__init__.py:
```diff
@@ -235,6 +236,30 @@ def _float_or_str_size(size):
+    **artist_kws,
+):
+    """Interface to Bokeh for a histogram bar plot."""
+    artist_kws.setdefault("level", "glyph")
```
is this necessary?

OriolAbril commented on src/arviz_plots/plots/distplot.py:
```diff
@@ -178,6 +181,8 @@ def plot_dist(
         kind = rcParams["plot.density_kind"]
     if plot_kwargs is None:
         plot_kwargs = {}
+    if kind in ("hist", "ecdf"):
```
```suggestion
    if kind in ("kde", "dot"):
```

OriolAbril commented on src/arviz_plots/plots/utils.py:
```diff
@@ -87,3 +89,56 @@ def filter_aes(pc, aes_map, artist, sample_dims):
+
+
+# WIP: dropping bin_midpoints calculation and keeping bin_edges
+def restructure_hist_data(hist_dict):
```
this can be removed

OriolAbril commented:
This will need to be rebased and double checked to make it compatible with plotly. Take care of these last two comments left and after that I'll rebase and add more commits to the PR so it works with plotly
